### PR TITLE
Orb slam2 win

### PR DESCRIPTION
**Pull request by ghost**
update cmakelist.txt with will call thirdparty lib(g2o and DBoW2) cmakelist.txt so that it will avoid individual build process separately.
four options are added which can be called with cmake with ON or OFF option value to customize orb-slam2 project. by default it is shared library
a. ORBSKAM2_STATIC_LIB (for making orb-slam2 project as static or shared library)
b. G2O_STATIC_LIB (for making g2o as static or share library)
c. DBOW2_STATIC_LIB (for making Dbow2 as static or shared library)
removed usleep with std:thread::sleep_for()
for windows only static build is supported.
before building orb-slam2 project , opencv and pangolin should be present.

### CODE REVIEW

**Comment by paul-michalik** [comment]
So Ansuman, are you going to implement what we've just discussed? That would be:

- Add build_windows.bat, build_linux.sh using the similar semantics and structure as in Macs. 
```
Scenario: Build ORB_SLAM2 and examples according to options 
 Given the basic development for software development is set up 
 And the third party dependencies for ORB_SLAM2 are available to CMake find_package
 When I run build_<platform> <options>
 Then the ORB_SLAM2 project and examples (according to options) are built
 Examples:
 | options | default |
 | -DORBSLAM2_STATIC_LIB | off |
 | -DBUILD_EXAMPLES | off
```

Please check for completeness...

- Add bootstrap_windows.bat, bootstrap_linux.bat:

```
Given the basic development for software development is set up
When I run bootstrap_<platform>
Then the third party dependencies for ORB_SLAM2 are installed/made available
```

On Windows you can assume that there is a vcpkg installation on /Software/vcpkg-<tag> and run ``/Software/vcpkg-<tag>/vcpkg.exe  install opencv eigen3 --triplet ...``. See if that is feasible if not just set up the ``CMAKE_PREFIX_PATH`` as we do in Macs

**Comment by apattnaik0721013** [comment]
@paul-michalik  i have added 4 script:
if orb-slam dependency library already present, then following script will build and run orb-slam project
1.build_windows.bat [platform] [toolset] [Configuration] 
  ex: build_windows.bat [x64] [v141] [Release] 
  note: need to set ORB_3RD_PARTY_LIB_DIR variable with pre exist orb-slam dependency library path
2.build_linux.sh [platform] [toolset] [configuration]
  ex: build_linux.sh
if all the orb-slam dependecy need to installed, following script will handle automatically
3. bootstrap_windows.bat [platform] [toolset] [Configuration]
   ex: bootstrap_windows.bat x64 v141 Release
4. bootstrap_linux.bat [platform] [toolset] [Configuration]
  ex:bootstrap_linux.bat
  note: currently bootstrap_linux.bat will install library for ubuntu 
5. build.sh is unchanged which means it is kept as original

File: bootstrap_windows.bat [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+setlocal
+
+set "OrbSlamPlatform=x86"
+set "OrbSlamToolset=v140"
+set "Buildtype=Release"
+
+if NOT "%~1"=="" set "OrbSlamPlatform=%~1"
+if NOT "%~2"=="" set "OrbSlamToolset=%~2"
+if NOT "%~3"=="" set "Buildtype=%~3"
+
```
Comment (paul-michalik) [inline bootstrap_windows.bat]: I propose that we assume vcpkg is already installed. Equivalent to the assumption that apt-get is there...  So here we just skip the portion which attempt to checkout vcpkg and go straight towards using it... Much simpler and scales better with future plans.

**Comment by apattnaik0721013** [comment]
@paul-michalik i have address your comment and updated bootsrap_windows.bat.

**Comment by paul-michalik** [comment]
Looks good! One more thing: Please do not invoke build from bootstrap. It should be vice versa if at all. Let's keep the two scripts separate.

**Comment by apattnaik0721013** [comment]
@paul-michalik, ,as per your comment,i have made separate bootstap_windows.bat from build_windows.bat.

File: CMakeLists.txt [hunk 2]
```diff
@@ -45,10 +62,19 @@ ${PROJECT_SOURCE_DIR}/include
 ${EIGEN3_INCLUDE_DIR}
 ${Pangolin_INCLUDE_DIRS}
 )
-
+add_subdirectory(Thirdparty/DBoW2)
+add_subdirectory(Thirdparty/g2o)
 set(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${PROJECT_SOURCE_DIR}/lib)
```
Comment (paul-michalik) [inline CMakeLists.txt]: Remove, use standard build artifact location

File: CMakeLists.txt [hunk 3]
```diff
@@ -45,10 +62,19 @@ ${PROJECT_SOURCE_DIR}/include
...
+set(ORB_SLAM_G2O_LIBRARIES ${PROJECT_SOURCE_DIR}/Thirdparty/DBoW2/lib/libDBoW2.so)
+set(ORB_SLAM_DBoW2_LIBRARIES ${PROJECT_SOURCE_DIR}/Thirdparty/g2o/lib/libg2o.so)
 
-add_library(${PROJECT_NAME} SHARED
+IF(WIN32)
+set(ORB_SLAM_G2O_LIBRARIES ${PROJECT_SOURCE_DIR}/Thirdparty/g2o/bin/${CMAKE_BUILD_TYPE}/g2o.lib)
+set(ORB_SLAM_DBoW2_LIBRARIES ${PROJECT_SOURCE_DIR}/Thirdparty/DBoW2/lib/${CMAKE_BUILD_TYPE}/DBoW2.lib)
+set(ORM_THIRD_PARTY_LIB ${ORB_SLAM_G2O_LIBRARIES} ${ORB_SLAM_DBoW2_LIBRARIES})
+ENDIF()
+
```
Comment (paul-michalik) [inline CMakeLists.txt]: Remove all these custom variables. Use target names and `target_link_libraries` to refer to project targets

File: CMakeLists.txt [hunk 4]
```diff
@@ -74,40 +100,40 @@ target_link_libraries(${PROJECT_NAME}
...
-${PROJECT_SOURCE_DIR}/Thirdparty/DBoW2/lib/libDBoW2.so
-${PROJECT_SOURCE_DIR}/Thirdparty/g2o/lib/libg2o.so
+${ORB_SLAM_G2O_LIBRARIES}
+${ORB_SLAM_DBoW2_LIBRARIES}
 )
 
 # Build examples
-
+if(BUILD_EXAMPLES)
 set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${PROJECT_SOURCE_DIR}/Examples/RGB-D)
```
Comment (paul-michalik) [inline CMakeLists.txt]: Remove, use standard location for build artifacts

**Comment by paul-michalik** [comment]
yAnsuman, the following scenario does not work:

```
Given "bootstrap_windows.bat x86 v141 Debug" is run
When I build orbslam2.sln in Visual Studio 2017 or via msbuild per command line
Then the libraries and applications are built without errors
```
The Third party libraries are not found. I suggest following changes to the CMakeLists: Remove all the custom locations of build artifacts. This means all overrides of `CMAKE_LIBRARY_OUTPUT_DIRECTORY` and similar must be removed. Instead, standard target names should be used.
This will decouple the CMake definitions from the upstream but since they are broken anyway, who cares.... and we get a clean build definition.

**Review by paul-michalik — CHANGES_REQUESTED** [review]
Please follow the comments in code. The custom locations and custom names for targets must be removed.

**Comment by apattnaik0721013** [comment]
@paul-michalik 
i have some observation: can you give your input on this.
1. cmake variable is supported in cmake > 3.8 version.  
   e.g: target_compile_features(${PROJECT_NAME} PUBLIC cxx_std_11)
2. apt-get cmake latest version is cmake 5.4. but vcpkg in windows is 3.9
3. which means current code is working in windows and build failed in linux as cxx_std_11 variable unknown error is coming
for reference-: ( https://github.com/ceres-solver/ceres-solver/issues/273 )
Status completed:
1. made some modification to build on linux. (but with cmake 3.9.1 support. which is installed manually)
2. As apt-get install old opencv , currently latest opencv is downloaded and build from git as per the instruction provided by opencv build instruction.

**Comment by apattnaik0721013** [comment]
@paul-michalik
1. updated orb-slam2 dependencies for ubuntu
2. used lates opencv and cmake ,
3. set cmake minimum version is 3.8
4. add compiler flag -bigobj for building latest library as orb-slam2 in debug mode. other wise it will show error in debug mode.
5. add library package installation path

File: README.md [hunk 5]
```diff
@@ -88,6 +88,12 @@ chmod +x build.sh
 
 This will create **libORB_SLAM2.so**  at *lib* folder and the executables **mono_tum**, **mono_kitti**, **rgbd_tum**, **stereo_kitti**, **mono_euroc** and **stereo_euroc** in *Examples* folder.
 
+for Windows:
+we provide a script `build_win.bat` to build the *Thirdparty* libraries and *ORB-SLAM2*. Please make sure you have installed all required dependencies (see section 2). Execute:
+```
+cd ORB_SLAM2
+build_win.bat [platform] [toolsset] [configuration] (e.g-> build_win.bat x64 v141 Release)
+
```
Comment (paul-michalik) [inline README.md]: Please update the bootstrapping and build part of the readme!

**Review by paul-michalik — CHANGES_REQUESTED** [review]
The review tool is strange, comments disappear randomly.. Anyway please try to follow whenever possible.

File: bootstrap_linux.sh [hunk 6]
```diff
@@ -2,11 +2,14 @@
 
 wget -N https://cmake.org/files/v3.9/cmake-3.9.2-Linux-x86_64.sh -P /usr/
 chmod 755 /usr/cmake-3.9.2-Linux-x86_64.sh
-/usr/cmake-3.9.2-Linux-x86_64.sh --skip-license --prefix=/usr/
+mkdir -p /opt/cmake-3.9.2
+/usr/cmake-3.9.2-Linux-x86_64.sh --skip-license --prefix=/opt/cmake-3.9.2
+ln -s /opt/cmake-3.9.2/bin/cmake /usr/local/bin/cmake
```
Comment (paul-michalik) [inline bootstrap_linux.sh]: Please do not link this into /user/local/bin. I would only support a parallel installation which is **optionally** put into path if the user is willing to accept that. Just prepend it to path for current shell or invoke cmake using full path.

**Pull request merged**